designs/node-ownership.md

```diff
@@ -0,0 +1,227 @@
+
+### Race Conditions
+
+#### Instance Orphaning Due to Node Creation Conflict
+
+While Karpenter performs creation at the end of the provisioning loop, it is possible for the node to already exist at the Kubernetes api-server for an instance. In the unhappy path, if Karpenter is particularly slow, it is possible for the kubelet to register the Node prior to Karpenter (i.e. we have a large scale-up and the kubelet registers the node before all VM instances resolve their IPs). In this case, [Karpenter receives a conflict error](https://github.com/aws/karpenter-core/blob/842731178a3b732b8ef63f20f7ebc243f00d85fa/pkg/controllers/provisioning/provisioner.go#L352) from the api-server and assumes that the node was already registered by the kubelet for this VM instance.
+
+This becomes a problem specifically when the node object we receive a conflict for doesn’t actually represent the same node/instance registered by the kubelet. As an example, users have observed behavior where the VM instance is deleted in the background, without Karpenter’s knowledge, meaning that the node object for that instance still exists on the api-server. The node begins to report as unhealthy, pods are evicted and Karpenter launches additional capacity for the pods.
+
+Since Karpenter performs naming for nodes based on host naming (this is also the current default for EKS and can only be switched starting in [K8s v1.23](https://github.com/aws/containers-roadmap/issues/1723). Karpenter attempts to create the node for this instance; however, because the node will still exist for the old VM instance with the same node name, we will not perform node creation for this node, assuming that the kubelet has already registered it.
+
+**This is wrong, since it means that the node provider ID will map to the old instance and the termination finalizer logic will no longer handle deletion of the instance that Karpenter provisioned.**
+
+This issue is primarily caused by a few issues:
+
+1. cloud-controller-manager that is responsible for cleaning up nodes that no longer have instances tied to it does not have a quick enough reconciliation loop to cleanup the node before Karpenter provisions the new capacity. *NOTE: In general, we shouldn’t rely on this timing anyways since Kubernetes is eventually consistent.*
+2. The subnet is highly constrained leading to IP hostname naming conflicts with the node
+3. EKS nodes rely on hostnaming, which ultimately causes this conflict. This problem could theoretically be circumvented by using [resource name-based naming](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-naming.html) that would ensure each node has a unique name, regardless of IP.
```

Comment:
Perhaps we should explicitly recommend resource based naming (we can still support the other way).

Reply:
I don't think we can just rip out IP-based naming, particularly because it is only supported on a later version of AWS K8s that not all customers have migrated to, and also because just solving this doesn't cover the other problems that are outlined in the design.

Yes, we can recommend that users use this naming convention and/or use interruption handling to circumvent this race, but I think it's a non-starter to force everyone over to a naming convention.

Reply:
Yep, agreed. “Explicitly recommend” is different from ”insist upon”.